### ISSUE TITLE
Conan recipe to build C++ client and publication to conan-center

### ISSUE THREAD

**Issue by cbornet** [apache/pulsar#2990]
[Conan.io](https://conan.io/) is a popular C++ package manager. It would be very handy to have a conan recipe to build the C++ client. It would make it easy to build even on Windows, even with MSVC.
Ultimately, you could pre-build conan packages for various platforms and compilers so that it's easy to include the client lib in a conan-enabled project.

**Comment by cbornet** [comment]
I checked and all dependencies used by the C++ client seem to exist in conan except `log4cxx` (but it is optional, isn't it ?):
Boost, OpenSSL, zlib, json-cpp, curl, gtest/gmock --> conan-center
Boost-Python, protobuf,  --> bincrafters

Also the only Linux-only dependency seems to be the crc32 lib. Couldn't it be possible to use a cross-platform CRC32 so that the client can be built on Windows ?

**Comment by cbornet** [comment]
Related to #15 #2461 #2680

**Comment by merlimat** [comment]
@cbornet I don't have any experience with Conan.io but it seems to be a good way to package the client lib.

> log4cxx (but it is optional, isn't it ?): 

Yes, log4cxx is optional and disabled by default

> Boost-Python, protobuf --> bincrafters

boost-python is only needed at compile time (and we should refactor the CMakeList.txt to not require it at compile time as it's only used to build the python client wrapper).

> Also the only Linux-only dependency seems to be the crc32 lib. Couldn't it be possible to use a cross-platform CRC32 so that the client can be built on Windows ?

We don't depend on crc32 libraries. We have a crc32-c implementation imported in our sources.

**Comment by cbornet** [comment]
> We don't depend on crc32 libraries. We have a crc32-c implementation imported in our sources.

Yes. My point was more that this crc32-c implementation depends on pthread which is not natively supported on windows. Sorry for not being clear enough.
But there's actually a [conan package for pthread on windows](https://github.com/bincrafters/conan-pthread-win32) so maybe it could be used.

**Comment by cbornet** [comment]
Yes that's even better

**Comment by soroshsabz** [comment]
ITNOA

`__builtin_expect` does not exist in Windows on MSVC, did you any plan to resolve this?

**Comment by soroshsabz** [comment]
related to #3800
